File: examples/shipments/create_shipment.py
"""
Ejemplo: Crear Envío Completo con Skydropx API (Python)

Este ejemplo muestra el flujo completo:
1. Crear cotización
2. Esperar tarifas
3. Seleccionar mejor tarifa
4. Crear envío
5. Obtener etiqueta y número de rastreo
"""

import os
import sys
from pathlib import Path

# Agregar el directorio src al path
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / 'src' / 'clients' / 'python'))

from skydropx_client import SkydropxClient, SkydropxError
from dotenv import load_dotenv

# Cargar variables de entorno
load_dotenv()


def main():
    """Ejemplo de envío completo"""
    
    print('=' * 60)
    print('EJEMPLO: CREAR ENVÍO COMPLETO')
    print('=' * 60)
    print()
    
    try:
        # 1. Autenticar
        print('🔐 Autenticando...')
        client = SkydropxClient(
            client_id=os.getenv('SKYDROPX_CLIENT_ID'),
            client_secret=os.getenv('SKYDROPX_CLIENT_SECRET'),
            environment=os.getenv('SKYDROPX_ENVIRONMENT', 'sandbox')
        )
        client.authenticate()
        print('✅ Autenticado\n')
        
        # 2. Crear cotización
        print('💰 Paso 1: Creando cotización...')
        quotation = client.create_quotation({
            'address_from': {
                'country_code': 'MX',
                'postal_code': '64000',
                'area_level1': 'Nuevo León',
                'area_level2': 'Monterrey',
                'area_level3': 'Centro'
            },
            'address_to': {
                'country_code': 'MX',
                'postal_code': '01000',
                'area_level1': 'Ciudad de México',
                'area_level2': 'Álvaro Obregón',
                'area_level3': 'Santa Fe'
            },
            'packages': [
                {
                    'weight': 2.5,
                    'length': 30,
                    'width': 20,
                    'height': 15
                }
            ]
        })
        print(f"✅ Cotización creada: {quotation['id']}\n")
        
        # 3. Esperar tarifas
        print('⏳ Paso 2: Esperando tarifas', end='', flush=True)
        quotation_result = client.wait_for_quotation(quotation['id'])
        print(' ✅\n')
        
        # 4. Seleccionar mejor tarifa
        print('📊 Paso 3: Seleccionando mejor tarifa...')
        rates = [r for r in quotation_result.get('rates', []) if r.get('success')]
        
        if not rates:
            print('❌ No se encontraron tarifas disponibles')
            return
        
        # Ordenar por precio y seleccionar la más barata
        rates.sort(key=lambda r: float(r['total']))
        selected_rate = rates[0]
        
        print(f"✅ Tarifa seleccionada:")
        print(f"   Paquetería: {selected_rate['provider_display_name']}")
        print(f"   Servicio: {selected_rate['provider_service_name']}")
        print(f"   Precio: ${selected_rate['total']} {selected_rate['currency_code']}")
        print(f"   Días: {selected_rate['days']}")
        print(f"   Rate ID: {selected_rate['id']}")
        print()
        
        # 5. Crear envío
        print('📦 Paso 4: Creando envío...')
        shipment = client.create_shipment({
            'rate_id': selected_rate['id'],
            'printing_format': 'thermal',
            'address_from': {
                'name': 'Juan Pérez',
                'company': 'Mi Empresa SA de CV',
                'street1': 'Av. Constitución 123',
                'phone': '8112345678',
                'email': 'juan.perez@miempresa.com',
                'reference': 'Edificio azul, entrada principal'
            },
            'address_to': {
                'name': 'María García López',
                'company': 'Cliente Premium',
                'street1': 'Insurgentes Sur 456',
                'phone': '5587654321',
                'email': 'maria.garcia@cliente.com',
                'reference': 'Torre B, piso 3, tocar timbre'
            }
        })
        
        print('✅ ¡Envío creado exitosamente!\n')
        
        # 6. Mostrar información del envío
        shipment_data = shipment['data']
        print('📋 Información del envío:')
        print(f"   ID: {shipment_data['id']}")
        print(f"   Estado: {shipment_data['attributes']['workflow_status']}")
        print(f"   Total: ${shipment_data['attributes']['total']} {shipment_data['attributes']['currency']}")
        print()
        
        # 7. Obtener información del paquete
        package_info = None
        for item in shipment.get('included', []):
            if item['type'] == 'packages':
                package_info = item
                break
        
        if package_info:
            attrs = package_info['attributes']
            print('📦 Información del paquete:')
            print(f"   Número de rastreo: {attrs['tracking_number']}")
            print(f"   Estado: {attrs['tracking_status']}")
            print()
            
            if attrs.get('label_url'):
                print('🎫 Etiqueta de envío:')
                print(f"   {attrs['label_url']}")
                print('   ⬇️  Descarga e imprime esta etiqueta')
                print()
            
            if attrs.get('tracking_url_provider'):
                print('🔍 URL de rastreo:')
                print(f"   {attrs['tracking_url_provider']}")
                print()
        
        print('=' * 60)
        print('✅ ENVÍO COMPLETADO')
        print('=' * 60)
        print()
        print('📝 Próximos pasos:')
        print('   1. Descarga e imprime la etiqueta')
        print('   2. Pégala en tu paquete')
        print('   3. Programa una recolección o llévalo a sucursal')
        print()
        print('💡 Para rastrear el envío, usa:')
        print('   python examples/tracking/track_shipment.py')
        
    except SkydropxError as e:
        print()
        print('❌ Error de Skydropx:')
        print(f'   Mensaje: {e.message}')
        if e.status_code:
            print(f'   Código: {e.status_code}')
        if e.response_data:
            print(f'   Detalles: {e.response_data}')
        sys.exit(1)
        
    except Exception as e:
        print()
        print(f'❌ Error inesperado: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: examples/shipments/create-shipment.js
require('dotenv').config();
const SkydropxClient = require('../../src/clients/javascript/SkydropxClient');

/**
 * Ejemplo: Flujo completo de creación de envío
 * 
 * Este ejemplo muestra el flujo completo:
 * 1. Crear cotización
 * 2. Seleccionar mejor tarifa
 * 3. Crear envío con la tarifa seleccionada
 * 4. Obtener etiqueta de envío
 */

async function main() {
  console.log('📦 Ejemplo de Creación de Envío Completo\n');
  
  const client = new SkydropxClient({
    clientId: process.env.SKYDROPX_CLIENT_ID,
    clientSecret: process.env.SKYDROPX_CLIENT_SECRET,
    environment: process.env.SKYDROPX_ENVIRONMENT || 'sandbox'
  });
  
  try {
    // 1. Autenticar
    await client.authenticate();
    console.log('✅ Autenticado correctamente\n');
    
    // 2. Crear cotización
    console.log('📝 Paso 1: Creando cotización...');
    const quotation = await client.createQuotation({
      address_from: {
        country_code: 'MX',
        postal_code: '64000',
        area_level1: 'Nuevo León',
        area_level2: 'Monterrey',
        area_level3: 'Centro'
      },
      address_to: {
        country_code: 'MX',
        postal_code: '03100',
        area_level1: 'Ciudad de México',
        area_level2: 'Benito Juárez',
        area_level3: 'Del Valle Centro'
      },
      packages: [{
        weight: 1.5,
        length: 25,
        width: 15,
        height: 10
      }]
    });
    
    console.log(`✅ Cotización creada: ${quotation.id}`);
    
    // 3. Esperar a que complete
    console.log('\n⏳ Esperando tarifas...');
    let quotationResult = quotation;
    let attempts = 0;
    
    while (!quotationResult.is_completed && attempts < 15) {
      await sleep(2000);
      quotationResult = await client.getQuotation(quotation.id);
      attempts++;
      process.stdout.write('.');
    }
    console.log(' ✅\n');
    
    // 4. Seleccionar mejor tarifa
    console.log('📝 Paso 2: Seleccionando tarifa...');
    const successfulRates = quotationResult.rates.filter(r => r.success);
    
    if (successfulRates.length === 0) {
      console.log('❌ No hay tarifas disponibles');
      return;
    }
    
    // Seleccionar la más económica
    const selectedRate = successfulRates.sort((a, b) => 
      parseFloat(a.total) - parseFloat(b.total)
    )[0];
    
    console.log(`✅ Tarifa seleccionada: ${selectedRate.provider_display_name}`);
    console.log(`   Servicio: ${selectedRate.provider_service_name}`);
    console.log(`   Precio: $${selectedRate.total} ${selectedRate.currency_code}`);
    console.log(`   Días: ${selectedRate.days}`);
    
    // 5. Crear envío
    console.log('\n📝 Paso 3: Creando envío...');
    const shipment = await client.createShipment({
      rate_id: selectedRate.id,
      printing_format: 'thermal', // o 'standard'
      address_from: {
        name: 'Juan Pérez',
        company: 'Mi Empresa SA de CV',
        street1: 'Av. Constitución 123',
        phone: '8112345678',
        email: 'juan.perez@miempresa.com',
        reference: 'Edificio azul, entrada principal'
      },
      address_to: {
        name: 'María García López',
        company: 'Cliente Premium',
        street1: 'Calle Insurgentes 456 Int 302',
        phone: '5587654321',
        email: 'maria.garcia@cliente.com',
        reference: 'Torre B, piso 3'
      }
    });
    
    console.log('✅ Envío creado exitosamente!\n');
    
    // 6. Mostrar información del envío
    console.log('📊 Información del envío:');
    console.log(`   ID: ${shipment.data.id}`);
    console.log(`   Estado: ${shipment.data.attributes.workflow_status}`);
    console.log(`   Paquetería: ${shipment.data.attributes.carrier_name.toUpperCase()}`);
    console.log(`   Total: $${shipment.data.attributes.total}`);
    console.log(`   Estado de pago: ${shipment.data.attributes.payment_status}`);
    
    // 7. Información del paquete
    const packageInfo = shipment.included.find(item => item.type === 'packages');
    if (packageInfo) {
      console.log('\n📦 Información del paquete:');
      console.log(`   ID: ${packageInfo.id}`);
      console.log(`   Número de rastreo: ${packageInfo.attributes.tracking_number || 'Pendiente'}`);
      console.log(`   Estado: ${packageInfo.attributes.tracking_status || 'Creado'}`);
      console.log(`   Peso: ${packageInfo.attributes.weight} kg`);
      console.log(`   Dimensiones: ${packageInfo.attributes.length}x${packageInfo.attributes.width}x${packageInfo.attributes.height} cm`);
      
      if (packageInfo.attributes.label_url) {
        console.log(`\n🎫 Etiqueta de envío: ${packageInfo.attributes.label_url}`);
        console.log('   ⬇️  Descarga e imprime esta etiqueta para tu paquete');
      }
      
      if (packageInfo.attributes.tracking_url_provider) {
        console.log(`\n🔍 Rastreo: ${packageInfo.attributes.tracking_url_provider}`);
      }
    }
    
    // 8. Direcciones
    const addressFrom = shipment.included.find(item => 
      item.type === 'addresses' && item.attributes.address_type === 'from'
    );
    const addressTo = shipment.included.find(item => 
      item.type === 'addresses' && item.attributes.address_type === 'to'
    );
    
    if (addressFrom && addressTo) {
      console.log('\n📍 Direcciones:');
      console.log(`   Origen: ${addressFrom.attributes.name}, ${addressFrom.attributes.area_level2}, ${addressFrom.attributes.postal_code}`);
      console.log(`   Destino: ${addressTo.attributes.name}, ${addressTo.attributes.area_level2}, ${addressTo.attributes.postal_code}`);
    }
    
    console.log('\n✅ ¡Envío listo para ser recolectado o entregado en sucursal!');
    
  } catch (error) {
    console.error('❌ Error:', error.message);
    if (error.details) {
      console.error('   Detalles:', JSON.stringify(error.details.data, null, 2));
    }
    process.exit(1);
  }
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Ejecutar ejemplo
main();
